File: c++/字符串/Hash/hash.cpp
namespace HASH
{
	#define type int
	int m;
	const int sd0=131;
	const int sd1=233;
	const int p0=402653189;
	const int p1=805306457;
	vector<type> tmp[2];
	void init(int _m)
	{
		int i,j;
		m=_m;
		tmp[0].resize(m+1);
		tmp[1].resize(m+1);
		tmp[0][0]=tmp[1][0]=1;
		for(i=1;i<=m;i++)
		{
			tmp[0][i]=1LL*tmp[0][i-1]*sd0%p0;
			tmp[1][i]=1LL*tmp[1][i-1]*sd1%p1;
		}
	}
	struct hash_table
	{
		vector<type> ha[2];
		void work(char *s,int n)
		{
			int i,j;
			assert(n<=m);
			ha[0].resize(n+1);
			ha[1].resize(n+1);
			ha[0][0]=ha[1][0]=0;
			for(i=1;i<=n;i++)
			{
				ha[0][i]=(1LL*ha[0][i-1]*sd0+s[i]-'a'+1)%p0;
				ha[1][i]=(1LL*ha[1][i-1]*sd1+s[i]-'a'+1)%p1;
			}
		}
		vector<type> get(int l,int r)
		{
			vector<type> res(2);
			res[0]=(ha[0][r]-1LL*ha[0][l-1]*tmp[0][r-l+1])%p0;
			if(res[0]<0) res[0]+=p0;
			res[1]=(ha[1][r]-1LL*ha[1][l-1]*tmp[1][r-l+1])%p1;
			if(res[1]<0) res[1]+=p1;
			return res;
		}
	};
	#undef type
}
HASH::hash_table ha;
/*
HASH::init(m);
ha.work(s,n);   // n<=m
*/

File: kotlin/hash.kt
class hash_table
{
	var m=0
	var ha1=LongArray(1)
	var ha2=LongArray(1)
	var t1=LongArray(1)
	var t2=LongArray(1)
	val p1=1e9.toLong()+9L
	val p2=998244353L
	val sd1=131L
	val sd2=233L
	val offset=1e9.toLong()
	fun init(n:Int)
	{
		t1=LongArray(n+5)
		t2=LongArray(n+5)
		t1[0]=1L
		t2[0]=1L
		m=n
		for(i in 1..m)
		{
			t1[i]=t1[i-1]*sd1%p1
			t2[i]=t2[i-1]*sd2%p2
		}
	}
	fun work(s:String,n:Int)
	{
		require(m>=n)
		ha1=LongArray(n+5)
		ha2=LongArray(n+5)
		ha1[0]=0
		ha2[0]=0
		for(i in 1..n)
		{
			ha1[i]=(ha1[i-1]*sd1+s[i-1].toLong())%p1
			ha2[i]=(ha2[i-1]*sd2+s[i-1].toLong())%p2
		}
	}
	fun get(l:Int,r:Int):Long
	{
		var res1=(ha1[r]-ha1[l-1]*t1[r-l+1])%p1
		if(res1<0) res1+=p1
		var res2=(ha2[r]-ha2[l-1]*t2[r-l+1])%p2
		if(res2<0) res2+=p2
		var res=res1*offset+res2
		return res
	}
}
